### ISSUE TITLE
Allow direct import of glossaries, e.g. ISO document terms

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/metanorma#75]
There is a need to directly import terms from online-accessible glossaries.

```adoc
=== date

[.source]
<<ISO8601-1>>
```

With this bibliography:

```adoc
[bibliography]
== Bibliography

* [[[ISO8601-1,ISO 8601-1]]]
```

=> should automatically expand to the effect of:

```adoc
=== date

time on the calendar time scale

NOTE: Common forms of date include: calendar date ...
```

**Comment by opoudjis** [comment]
Will do IEV as an initial instance, not sure what other online glossaries for ISO are going to be accessible at all. Will require IEV API integration. Could use termbase invocation in source as means of triggering expansion.

**Comment by ronaldtse** [comment]
@opoudjis we can do Geolexica as a second one. That depends on https://github.com/geolexica/geolexica-server/issues/2 being completed.

**Comment by opoudjis** [comment]
@w00lf Propose to hand this over to you...

**Comment by opoudjis** [comment]
Concepts are already implemented: they are *references* to external termbases:

` {{blah,clause=3.1,figure=a,word}}` means "_Word_ is defined in document _blah_, clause 3.1, figure a"

`{{IEV:135-13-13,word,term}}` means "_term_ is defined as _word_ in IEV, 135-13-13."

But those are references. These are instead going to be importing the definition from that source.

So:

````
== Terms & definitions
=== date

[.source]
{{IEV:135-13-13,word,term}}
````

This is a term definition with a source but no body. I propose that in term cleanup, whenever we have a term definition with a concept source, and no definition:

````
xmldoc.xpath("//term").each do |x|
  if c = x.at("./source/concept") and !x.at("./definition")
    x.at("./source").previous = fetch_termbase(c["base"], c.text) 
  end
end
````

**Comment by opoudjis** [comment]
And that would live in:

metanorma-standoc/lib/asciidoc/standoc/cleanup_section.rb : termdef_cleanup()

You would call it as a separate new method, before all the other methods called by termdef_cleanup().

**Comment by ronaldtse** [comment]
> we can do Geolexica as a second one. That depends on geolexica/geolexica-server#2 being completed

Is completed now.

**Comment by opoudjis** [comment]
Note that we also have 

````
== Terms & definitions
=== date

definition

[.source]
{{IEV:135-13-13,word,term}}, with adjustments
````

@w00lf you will not process this case, because a definition is supplied. But I need to process it, to ensure that the "modifications" are processed with a source that is a concept reference. So this ticket is temporarily my work as well.

**Comment by ronaldtse** [comment]
Just to confirm we need to differentiate these use cases:

Referring to a Concept ID within definition or text:

```adoc
=== date

definition blah {{IEV:1234,clause=1234,foo}}
```

Sourcing an entire Concept to the document's T&D:

```adoc
=== date

[.source]
{{IEV:1234,clause=1234}}
```

Sourcing an entire Concept to the document's T&D and also modifying its definition:

```adoc
=== date

blah blah

[.source]
{{IEV:1234,clause=1234}}, definition changed to "..."
```

**Comment by opoudjis** [comment]
Implementing third use case

**Comment by opoudjis** [comment]
I've added the stub for you, @w00lf, `def fetch_termbase()` in metanorma-standoc/lib/asciidoctor/standoc/cleanup_ref.rb

We expect this to be the contents of `<term>`. Term has the grammar:

````
term =
  element term {
    attribute id { xsd:ID }?,
    preferred+, admitted*, related*, deprecates*,
    termdomain?, termgrammar*, definition, termnote*, termexample*, termsource*
  }
````

We provide `term`, `term/@id`, `term/preferred`, and `termsource`. We inject the contents of fetch_termbase() before termsource; we expect that to be a nodeset of 

`admitted*, related*, deprecates*,
    termdomain?, termgrammar*, definition, termnote*, termexample*`

See https://github.com/metanorma/metanorma-model-iso/blob/master/grammars/isodoc.rnc

**Comment by ronaldtse** [comment]
@w00lf please also check out the concept model used by the glossarist sites: https://github.com/metanorma/metanorma-standoc/issues/255

Thanks!

**Comment by opoudjis** [comment]
Yes: this task is now blocked by metanorma/metanorma-standoc#255. I need to work out a serialisation of that information model into Metanorma XML, overriding the current one; and your target output is that new serialisation.

**Comment by ronaldtse** [comment]
Related to https://github.com/glossarist/glossarist-ruby . 

Currently we only support sourcing IEV terms.

We can import from additional termbases for:
* import data of a term + provide bibliographic referencing for sourcing (using the glossarist gem)
* just provide bibliographic referencing

Full termbase sources to import from:
* Geolexica network sites (https://isotc211.geolexica.org, https://isotc204.geolexica.org, ...)
* Geolexica client sites (https://glossary.metaverse-standards.org)
* Glossarist datasets on GitHub
* IEV (right now we only support bib referencing, not the data)

Bibliographic termbase sources:
* ISO OBP
* ITU Glossary
* UN glossary

RDF sources:
* OGC Definitions Server

**Comment by opoudjis** [comment]
Pending https://github.com/geolexica/jekyll-geolexica/issues/14

Once that is actioned, Geolexica.org will be the initial instance for a generic glossarist-base implementation of reading termbases.

**Comment by opoudjis** [comment]
Geolexica has been implemented.

**Comment by opoudjis** [comment]
@HassanAkbar  I need help here, as I've alerted you.

I am about to test the ability of lutaml-plugin-glossarist to import definitions from a termbase, and incorporate them into a Metanorma document. This will be tested based on Geolexica.

I have no idea what has been happening and what is involved. I gather you do? Could you please point me to any documentation of how this is done?

**Comment by HassanAkbar** [comment]
@opoudjis I can't remember any gem named `lutaml-plugin-glossarist` do you mean `metanorma-plugin-glossarist`?

If yes then currently the only documentation for that is the [GitHub readme](https://github.com/metanorma/metanorma-plugin-glossarist/blob/main/README.md) or you can take a look at the specs for [some examples here](https://github.com/metanorma/metanorma-plugin-glossarist/blob/main/spec/metanorma/plugin/glossarist/dataset_preprocessor_spec.rb).

If there is some confusion just let me know and I'll try to clear that and also update that in the readme for future use.

**Comment by opoudjis** [comment]
@HassanAkbar What I need to understand is how to hook metanorma-plugin-glossarist up to Geolexica.org: how do I (or preferably not I) export or get geolexica.org into a local file, that I can manipulate with `:glossarist-dataset: dataset:./path/to/glossarist-dataset` ? Is it just https://github.com/geolexica/jekyll-geolexica/blob/main/_data/lang.yaml ?

**Comment by HassanAkbar** [comment]
@opoudjis Here are the datasets for the geolexica sites,
- [isotc211-glossary](https://github.com/geolexica/isotc211-glossary) for [isotc211.geolexica.org](https://isotc211.geolexica.org)
- [isotc204-glossary](https://github.com/geolexica/isotc204-glossary) for [isotc204.geolexica.org](https://isotc204.geolexica.org)

`./path/to/glossarist-dataset` should be `isotc211-glossary/geolexica` for `isotc211-glossary` and `isotc204-glossary/concepts` for `isotc204-glossary`

**Comment by opoudjis** [comment]
Found two issues with Glossarist on attempting to do the basic import of the entire termbase. There will be others, but this issue for now goes on hold until they are addressed.

**Comment by opoudjis** [comment]
Got TC 211 compiling. There are lots of issues.

**Comment by opoudjis** [comment]
Improving reporting of duplicate terms, because I'm having trouble finding them.
